### DISCUSSION TITLE
Launching a service

### DISCUSSION THREAD

**Discussion by meshvaD** [uwrobotics/uwrt_mars_rover#166] [Q&A]
I'm trying to launch a service (for the last part of the training), and it shows up in the "service list" . But, whenever I try to call it, it says "waiting for service to become available...". 

Right now, I just created a launch_ros.actions.Node for the service, but is there something else you need to do to launch the it?
I was following along with this to make the file: https://docs.ros.org/en/foxy/Tutorials/Launch-Files/Creating-Launch-Files.html

**Comment by wmmc88** [comment]
Which service are you trying to call? the one that you wrote? did you setup the service server properly in ur node code? could you post a code snippet?

**Comment by meshvaD** [comment]
```
class SpawnTurtle : public rclcpp::Node
{
public:
    explicit SpawnTurtle(const rclcpp::NodeOptions &options)
        : Node("spawn_turtle")
    {
        this->srv_client = this->create_client<turtlesim::srv::Spawn>("spawn");
        this->timer = this->create_wall_timer(std::chrono::milliseconds(50), 
            std::bind(&SpawnTurtle::spawn_turtles, this));
    }
private:
    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr srv_client;
    rclcpp::TimerBase::SharedPtr timer;
    void spawn_turtles()
    {
        std::map<std::string, std::vector<int>>::iterator it;
        for (it = turtles.begin(); it != turtles.end(); it++)
        {
            std::unique_ptr<turtlesim::srv::Spawn_Request> request = 
                std::make_unique<turtlesim::srv::Spawn_Request>();

            request->name = it->first;
            request->x = it->second[0];
            request->y = it->second[1];

            auto end = [this] (rclcpp::Client<turtlesim::srv::Spawn>::SharedFuture response)
            -> void {
                rclcpp::shutdown();
            };

            this->srv_client->async_send_request(std::move(request), end);
        }
    }
};
}
RCLCPP_COMPONENTS_REGISTER_NODE(composition::SpawnTurtle)
```

**Reply by niiquaye** [reply]
Taking a quick look at your code it seems mostly correct from a quick run through (I recommend separating your code into header and source files - good practice).  Try double checking you are calling the correct service make sure the name is exactly as you see when you run a ros2 service show in the terminal - from the looks of things its either A) your client is calling a service that does not exist or has not been started up yet or B) you misspelled the name of the service. Double check you have the correct name of the service. Typically service names start with a '/<name of service>' .i.e. check that the service name is "/spawn" instead of just "spawn"

**Reply by meshvaD** [reply]
I tried that but it still isn't working.

**Reply by niiquaye** [reply]
Can you ping me on dicsord please. Lets talk more about this.

**Reply by niiquaye** [reply]
> Taking a quick look at your code it seems mostly correct from a quick run through (I recommend separating your code into header and source files - good practice). Try double checking you are calling the correct service make sure the name is exactly as you see when you run a ros2 service show in the terminal - from the looks of things its either A) your client is calling a service that does not exist or has not been started up yet or B) you misspelled the name of the service. Double check you have the correct name of the service. Typically service names start with a '/' .i.e. check that the service name is "/spawn" instead of just "spawn"

You are missing your visibility header stuff - AKA - macros that need to be defined before the function to help with linking of shared libraries. Take a closer look at the example code I have provided. You have followed it pretty nicely, but there are a few subtle differences. Also ensure that your class and namespace names & spelling are correct. I cant see from this snippet. Please ping on discord.
